README.md
# generate-robotstxt

[![NPM version](https://img.shields.io/npm/v/generate-robotstxt.svg)](https://www.npmjs.org/package/generate-robotstxt)
[![Travis Build Status](https://img.shields.io/travis/itgalaxy/generate-robotstxt/master.svg?label=build)](https://travis-ci.org/itgalaxy/generate-robotstxt)
[![dependencies Status](https://david-dm.org/itgalaxy/generate-robotstxt/status.svg)](https://david-dm.org/itgalaxy/generate-robotstxt)
[![devDependencies Status](https://david-dm.org/itgalaxy/generate-robotstxt/dev-status.svg)](https://david-dm.org/itgalaxy/generate-robotstxt?type=dev)

Awesome generator robots.txt.

## Installation

```shell
npm install --save-dev generate-robotstxt
```

## Usage

```js
import robotstxt from "generate-robotstxt";

robotstxt({
  policy: [
    {
      userAgent: "Googlebot",
      allow: "/",
      disallow: "/search",
      crawlDelay: 2
    },
    {
      userAgent: "OtherBot",
      allow: ["/allow-for-all-bots", "/allow-only-for-other-bot"],
      disallow: ["/admin", "/login"],
      crawlDelay: 2
    },
    {
      userAgent: "*",
      allow: "/",
      disallow: "/search",
      crawlDelay: 10,
      cleanParam: "ref /articles/"
    }
  ],
  sitemap: "http://example.com/sitemap.xml",
  host: "http://example.com"
}).then(content => {
  console.log(content);
});
```

## File based configuration

**robots-txt.config.js**

```js
module.exports = {
  policy: [
    {
      userAgent: "Googlebot",
      allow: "/",
      disallow: ["/search"],
      crawlDelay: 2
    },
    {
      userAgent: "OtherBot",
      allow: ["/allow-for-all-bots", "/allow-only-for-other-bot"],
      disallow: ["/admin", "/login"],
      crawlDelay: 2
    },
    {
      userAgent: "*",
      allow: "/",
      disallow: "/search",
      crawlDelay: 10,
      cleanParam: "ref /articles/"
    }
  ],
  sitemap: "http://example.com/sitemap.xml",
  host: "http://example.com"
};
```

## CLI

```shell
Awesome generator robots.txt

  Usage generate-robotstxt [options] <dest>

  Options:
     --config  Path to a specific configuration file.
```

## Contribution

Feel free to push your code if you agree with publishing under the MIT license.

## [Changelog](CHANGELOG.md)

## [License](LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [8.0.2](https://github.com/itgalaxy/generate-robotstxt/compare/v8.0.1...v8.0.2) (2020-03-24)

### Chore

* update deps

## [8.0.1](https://github.com/itgalaxy/generate-robotstxt/compare/v8.0.0...v8.0.1) (2020-02-19)

### Chore

* update deps

## [8.0.0](https://github.com/itgalaxy/generate-robotstxt/compare/v7.1.0...v8.0.0) (2019-07-03)

### Breaking change

* minimum require Node.js version is `10.13.0`.

## [7.1.0](https://github.com/itgalaxy/generate-robotstxt/compare/v6.0.1...v7.1.0) (2019-07-03)


### Features

* cjs export ([#123](https://github.com/itgalaxy/generate-robotstxt/issues/123)) ([9c0507b](https://github.com/itgalaxy/generate-robotstxt/commit/9c0507b))



## 7.0.0 - 2019-07-03

- Changed: minimum require Node.js version is `8.9.0`.

## 6.0.1 - 2019-03-26

- Chore: migrate on `ip-regex` to avoid lodash vulnerable libraries

## 6.0.0 - 2019-01-09

- Chore: minimum required `node-fs` version is now `^7.0.1`.

### BREAKING CHANGE

- Chore: minimum required `nodejs` version is now `>= 6.9.0`.

## 5.0.7 - 2018-05-22

- Chore: minimum required `cosmiconfig` version is now `^5.0.0`.

## 5.0.6 - 2018-05-05

- Fix: throw error when `policy` option is `null`.
- Chore: minimum required `meow` version is now `^5.0.0`.
- Chore: minimum required `node-fs` version is now `^6.0.0`.

## 5.0.5 - 2018-03-28

- Fix: allow empty values `Disallow` directive (according [official spec](http://www.robotstxt.org/orig.html)).
- Fix: don't generate empty `Clean-param`.

## 5.0.4 - 2018-01-16

- Fix: typo in source code message errors.

## 5.0.3 - 2018-01-16

- Chore: minimum required `cosmiconfig` version is now `^4.0.0`.

## 5.0.2 - 2017-12-12

- Chore: minimum required `node-fs` version is now `^5.0.0`.

## 5.0.1 - 2017-11-28

- Chore: minimum required `meow` version is now `^4.0.0`.

## 5.0.0 - 2017-11-15

- Changed: use `[cosmiconfig](https://github.com/davidtheclark/cosmiconfig) for
  loading configuration.
- Feature: in CLI if the parent directory does not exist when you write `robots.txt`, it's created.

## 4.0.4 - 2017-10-09

- Chore: update deps.

## 4.0.3 - 2017-03-13

- Fixed: `is-absolute-url` package semver.

## 4.0.2 - 2016-12-30

- Fixed: `host` options is now processed based `URL`.
- Fixed: thrown error if the `host` option being IP address.
- Fixed: clarified error message on multiple and not string the `userAgent`
  option.
- Fixed: `Host` directive is now not containing `80` port.
- Fixed: thrown error if the `cleanParam` not string or array and if string not
  more than 500 characters.
- Fixed: supported unicode characters in a `Allow` and a `Disallow` directives.
- Fixed: thrown error if the `sitemap` option not an array or a string and not
  an absolute URL.

## 4.0.1 - 2016-10-27

- Chore: added CI test on `node.js` version `7`.
- Documentation: improve `README.md` and fix typos.

## 4.0.0

- Added: `crawlDelay` to each `police` item.
- Added: `cleanParam` to each `police` item (used only Yandex bot).
- Chore: used `remark-preset-lint-itgalaxy` preset.
- Chore: updated `devDependencies`.
- Chore: updated copyright year in `LICENSE`.
- Chore: improved tests.
- Fixed: strict order directives for each `User-agent`.
- Fixed: added newline after each `User-agent`.
- Removed: `crawlDelay` from `options`.
- Removed: `cleanParam` from `options`.


test/index.js
/* global describe, it */
var should = require('should');
var generateRobotstxt = require('../index.js');
var fs = require('fs');

describe('Robots.txt', function () {
    it('should throw error if callback not function', function () {
        generateRobotstxt.bind(null, { }, null).should.throw('Callback must be function');
    });

    it('should call callback after generate and contents not empty without params', function (done) {
        generateRobotstxt(null, function (error, contents) {
            if (error) {
                return done(error);
            }

            should(contents).not.empty();

            return done();
        });
    });

    it('should contain one default policy without params', function (done) {
        generateRobotstxt({}, function (error, contents) {
            if (error) {
                return done(error);
            }

            should(contents).be.equal('User-agent: *\nAllow: /\n');

            return done();
        });
    });

    it('should contain two policy without params', function (done) {
        generateRobotstxt({
            policy: [
                {
                    userAgent: 'Google',
                    allow: '/'
                },
                {
                    userAgent: 'Yandex',
                    allow: '/'
                }
            ]
        }, function (error, contents) {
            if (error) {
                return done(error);
            }

            should(contents).be.equal('User-agent: Google\nAllow: /\nUser-agent: Yandex\nAllow: /\n');

            return done();
        });
    });

    it('should contain one default policy with Sitemap', function (done) {
        generateRobotstxt({
            sitemap: 'sitemap.xml'
        }, function (error, contents) {
            if (error) {
                return done(error);
            }

            should(contents).be.equal('User-agent: *\nAllow: /\nSitemap: sitemap.xml\n');

            return done();
        });
    });

    it('should contain one default policy with two Sitemaps', function (done) {
        generateRobotstxt({
            sitemap: [
                'sitemap.xml',
                'sitemap1.xml'
            ]
        }, function (error, contents) {
            if (error) {
                return done(error);
            }

            should(contents).be.equal('User-agent: *\nAllow: /\nSitemap: sitemap.xml\nSitemap: sitemap1.xml\n');

            return done();
        });
    });

    it('should contain one default policy with Host', function (done) {
        generateRobotstxt({
            host: 'http://domain.com'
        }, function (error, contents) {
            if (error) {
                return done(error);
            }

            should(contents).be.equal('User-agent: *\nAllow: /\nHost: http://domain.com\n');

            return done();
        });
    });

    it('should throw error with array of Host', function () {
        generateRobotstxt.bind(null, {
            host: [
                'http://domain.com',
                'http://domain1.com'
            ]
        }, function () {}).should.throw('Options Host must be one in robots.txt');
    });

    it('should contain one default policy with Craw-delay', function (done) {
        generateRobotstxt({
            crawDelay: 10
        }, function (error, contents) {
            if (error) {
                return done(error);
            }

            should(contents).be.equal('User-agent: *\nAllow: /\nCrawl-delay: 10\n');

            return done();
        });
    });

    it('should contain one default policy with one Clean-param', function (done) {
        generateRobotstxt({
            cleanParam: 's /forum/showthread.php'
        }, function (error, contents) {
            if (error) {
                return done(error);
            }

            should(contents).be.equal('User-agent: *\nAllow: /\nClean-param: s /forum/showthread.php\n');

            return done();
        });
    });

    it('should contain one default policy with two Clean-param', function (done) {
        generateRobotstxt({
            cleanParam: [
                's /forum/showthread.php',
                'ref /forum/showthread.php'
            ]
        }, function (error, contents) {
            if (error) {
                return done(error);
            }

            should(contents).be.equal('User-agent: *\nAllow: /\n'
                + 'Clean-param: s /forum/showthread.php\n'
                + 'Clean-param: ref /forum/showthread.php\n');

            return done();
        });
    });

    it('should throw error if Craw-delay not number', function () {
        generateRobotstxt.bind(null, {
            crawDelay: 'notNumber'
        }, function () {}).should.throw('Options Craw-delay must be integer or float');
    });

    it('should save robots.txt in destination', function (done) {
        var dest = './test/fixtures';

        generateRobotstxt({
            dest: dest
        }, function (robotsTxtError) {
            if (robotsTxtError) {
                return done(robotsTxtError);
            }

            fs.stat(dest, function (fsStatError) {
                if (fsStatError === null) {
                    return done();
                } else if (fsStatError.code === 'ENOENT') {
                    return done(new Error('File robots.txt not exist'));
                }

                return done(fsStatError);
            });
        });
    });

    it('should throw error if destination not exist', function (done) {
        generateRobotstxt({
            dest: '/test/test/test'
        }, function (error) {
            error.should.be.an.instanceof(Error);

            return done();
        });
    });

    it('should throw error if destination not string or null', function () {
        generateRobotstxt.bind(null, {
            dest: []
        }, function () {}).should.throw('Destination must be string or null');
    });
});


src/__tests__/index.js
import generateRobotstxt from '../index';
import path from 'path';
// eslint-disable-next-line node/no-unpublished-import
import test from 'ava';

const fixturesPath = path.join(__dirname, 'fixtures');

test('should generated default output without options',
    (t) => generateRobotstxt()
        .then((content) => {
            t.is(content, 'User-agent: *\nAllow: /\n');
        })
);

test('should contain two policy item with "Allow"',
    (t) => generateRobotstxt({
        policy: [{
            allow: '/',
            userAgent: 'Google'
        }, {
            allow: '/',
            userAgent: 'Yandex'
        }]
    })
        .then((content) => {
            t.is(
                content,
                'User-agent: Google\nAllow: /\n\n'
                    + 'User-agent: Yandex\nAllow: /\n'
            );
        })
);

test('should contain two policy item with "Allow" and "Disallow"',
    (t) => generateRobotstxt({
        policy: [{
            allow: '/',
            disallow: '/search-foo',
            userAgent: 'Google'
        }, {
            allow: '/',
            disallow: '/search-bar',
            userAgent: 'Yandex'
        }]
    })
        .then((content) => {
            t.is(
                content,
                'User-agent: Google\nAllow: /\nDisallow: /search-foo\n\n'
                    + 'User-agent: Yandex\nAllow: /\nDisallow: /search-bar\n'
            );
        })
);

test('should contain two policy item, first have multiple "User-agent"',
    (t) => generateRobotstxt({
        policy: [{
            allow: '/',
            disallow: '/search-foo',
            userAgent: ['Google', 'AnotherBot']
        }, {
            allow: '/',
            disallow: '/search-bar',
            userAgent: 'Yandex'
        }]
    })
        .then((content) => {
            t.is(
                content,
                'User-agent: Google\nUser-agent: AnotherBot\nAllow: /\nDisallow: /search-foo\n\n'
                    + 'User-agent: Yandex\nAllow: /\nDisallow: /search-bar\n'
            );
        })
);

test('should throw error if "policy" option is not array', (t) => {
    t.throws(generateRobotstxt({
        policy: 'string'
    }), 'Options "policy" must be array');
});

test('should throw error if "policy" option not have "userAgent"', (t) => {
    t.throws(generateRobotstxt({
        policy: [{}]
    }), 'Each "police" should have "User-agent"');
});

test('should throw error if "policy" option have empty "userAgent"', (t) => {
    t.throws(generateRobotstxt({
        policy: [{
            userAgent: []
        }]
    }), 'Each "police" should have "User-agent"');
});

test('should contain "Sitemap"',
    (t) => generateRobotstxt({
        sitemap: 'sitemap.xml'
    })
        .then((content) => {
            t.is(content, 'User-agent: *\nAllow: /\nSitemap: sitemap.xml\n');
        })
);

test('should contain two "Sitemap"',
    (t) => generateRobotstxt({
        sitemap: [
            'sitemap.xml',
            'sitemap1.xml'
        ]
    })
        .then((content) => {
            t.is(content, 'User-agent: *\nAllow: /\nSitemap: sitemap.xml\nSitemap: sitemap1.xml\n');
        })
);

test('should contain "Host"',
    (t) => generateRobotstxt({
        host: 'http://domain.com'
    })
        .then((content) => {
            t.is(content, 'User-agent: *\nAllow: /\nHost: http://domain.com\n');
        })
);

test('should throw error if "Host" option is array', (t) => {
    t.throws(generateRobotstxt({
        host: [
            'http://domain.com',
            'http://domain1.com'
        ]
    }), 'Options "host" must be one');
});

test('should contain multiple "User-agent" and "Crawl-delay"',
    (t) => generateRobotstxt({
        policy: [{
            allow: '/',
            crawlDelay: 10,
            userAgent: 'Google'
        }, {
            allow: '/',
            crawlDelay: 0.5,
            userAgent: 'Yandex'
        }]
    })
        .then((content) => {
            t.is(
                content,
                'User-agent: Google\nAllow: /\nCrawl-delay: 10\n\nUser-agent: Yandex\nAllow: /\nCrawl-delay: 0.5\n'
            );
        })
);

test('should throw error on invalid "crawlDelay" option', (t) => {
    t.throws(generateRobotstxt({
        policy: [{
            allow: '/',
            crawlDelay: 'foo',
            userAgent: 'Google'
        }]
    }), 'Options "crawlDelay" must be integer or float');
});

test('should contain one policy item with one "Clean-param"',
    (t) => generateRobotstxt({
        policy: [{
            allow: '/',
            cleanParam: 's /forum/showthread.php',
            userAgent: 'Yandex'
        }]
    })
        .then((content) => {
            t.is(content, 'User-agent: Yandex\nAllow: /\nClean-param: s /forum/showthread.php\n');
        })
);

test('should contain one policy item with two "Clean-params"',
    (t) => generateRobotstxt({
        policy: [{
            allow: '/',
            cleanParam: [
                's /forum/showthread.php',
                'ref /forum/showthread.php'
            ],
            userAgent: 'Yandex'
        }]
    })
        .then((content) => {
            t.is(
                content,
                'User-agent: Yandex\nAllow: /\n'
                    + 'Clean-param: s /forum/showthread.php\n'
                    + 'Clean-param: ref /forum/showthread.php\n'
            );
        })
);

test('should load config file',
    (t) => generateRobotstxt({
        configFile: `${fixturesPath}/config.js`
    })
        .then((content) => {
            t.is(
                content,
                'User-agent: *\nAllow: /\nHost: http://some-domain.com\n'
            );
        })
);

test('should load commonjs config file',
    (t) => generateRobotstxt({
        configFile: `${fixturesPath}/config-commonjs.js`
    })
        .then((content) => {
            t.is(
                content,
                'User-agent: *\nAllow: /\nHost: http://some-some-domain.com\n'
            );

            return content;
        })
);


src/cli.js
#!/usr/bin/env node

import path from "path";
import fs from "fs-extra";
import meow from "meow";
import resolveFrom from "resolve-from";
import standalone from "./standalone.js";

const cli = meow(
  `
    Usage generate-robotstxt [options] <dest>

    Options:
       --config  Path to a specific configuration file.
`,
  {
    flags: {
      config: {
        type: "string",
      },
    },
  }
);

const optionsBase = {};

if (cli.flags.config) {
  // Should check these possibilities:
  //   a. name of a node_module
  //   b. absolute path
  //   c. relative path relative to `process.cwd()`.
  // If none of the above work, we'll try a relative path starting
  // in `process.cwd()`.
  optionsBase.configFile =
    resolveFrom(process.cwd(), cli.flags.config) ||
    path.join(process.cwd(), cli.flags.config);
}

Promise.resolve()
  .then(() => Object.assign({}, optionsBase))
  .then((options) => standalone(options))
  .then((output) => {
    if (cli.input.length === 0) {
      throw new Error("Require `dest` argument");
    }

    const dest = path.resolve(cli.input.pop());

    return Promise.resolve().then(() => fs.outputFile(dest, output));
  })
  .catch((error) => {
    console.log(error); // eslint-disable-line no-console
    process.exit(error.code || 1);
  });


.eslintrc.js
"use strict";

module.exports = {
  extends: [
    "plugin:itgalaxy/script",
    "plugin:itgalaxy/esnext",
    "plugin:itgalaxy/node",
  ],
  overrides: [
    // Source
    {
      extends: ["plugin:itgalaxy/module"],
      // Exclude nested tests
      excludedFiles: ["**/__tests__/**/*", "**/__mocks__/**/*", "**/*.md"],
      files: ["src/**/*"],
      rules: {
        // Allow to use ES module syntax
        // You should use babel if your node version is not supported ES syntax module, dynamic loading ES modules or other features
        "node/no-unsupported-features/es-syntax": [
          "error",
          { ignores: ["modules", "dynamicImport"] },
        ],
      },
    },

    // Jest
    {
      extends: ["plugin:itgalaxy/dirty", "plugin:itgalaxy/jest"],
      excludedFiles: ["**/*.md"],
      files: ["**/__tests__/**/*", "**/__mocks__/**/*"],
      rules: {
        // Test can be written with using ES module syntax or CommonJS module syntax
        "node/no-unsupported-features/es-syntax": [
          "error",
          { ignores: ["modules", "dynamicImport"] },
        ],

        // Allow to use `console` (example - `mocking`)
        "no-console": "off",
      },
    },

    // Markdown
    {
      extends: [
        // Documentation files can contain ECMA and CommonJS modules
        "plugin:itgalaxy/dirty",
        "plugin:itgalaxy/markdown",
      ],
      files: ["**/*.md"],
      rules: {
        "no-unused-vars": "off",
        "no-console": "off",
        "import/no-unresolved": "off",
        "node/no-unpublished-require": "off",
        "node/no-unpublished-import": "off",
        // Documentation files can contain ES module syntax and CommonJS module syntax
        "node/no-unsupported-features/es-syntax": [
          "error",
          { ignores: ["modules", "dynamicImport"] },
        ],
      },
    },
  ],
  root: true,
};


src/standalone.js
import path from "path";
import url from "url";
import ipRegex from "ip-regex";
import { cosmiconfig } from "cosmiconfig";
import isAbsoluteUrl from "is-absolute-url";

function capitaliseFirstLetter(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}

function addLine(name, rule) {
  let contents = "";

  if (rule && Array.isArray(rule) && rule.length > 0) {
    rule.forEach((item) => {
      contents += addLine(name, item);
    });
  } else {
    const ruleContent = (name === "Allow" || name === "Disallow"
      ? encodeURI(rule)
      : rule
    ).toString();

    contents += `${capitaliseFirstLetter(
      name.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase()
    )}:${ruleContent.length > 0 ? ` ${ruleContent}` : ""}\n`;
  }

  return contents;
}

function generatePoliceItem(item, index) {
  let contents = "";

  if (index !== 0) {
    contents += "\n";
  }

  contents += addLine("User-agent", item.userAgent);

  if (item.allow) {
    contents += addLine("Allow", item.allow);
  }

  if (typeof item.disallow === "string" || Array.isArray(item.disallow)) {
    contents += addLine("Disallow", item.disallow);
  }

  if (item.crawlDelay) {
    contents += addLine("Crawl-delay", item.crawlDelay);
  }

  // Move from policy for next master version
  // https://yandex.ru/support/webmaster/controlling-robot/robots-txt.html
  if (item.cleanParam && item.cleanParam.length > 0) {
    contents += addLine("Clean-param", item.cleanParam);
  }

  return contents;
}

function buildConfig(configFile = null) {
  let searchPath = process.cwd();
  let configPath = null;

  if (configFile) {
    searchPath = null;
    configPath = path.resolve(process.cwd(), configFile);
  }

  const configExplorer = cosmiconfig("robots-txt");
  const searchForConfig = configPath
    ? configExplorer.load(configPath)
    : configExplorer.search(searchPath);

  return searchForConfig.then((result) => {
    if (!result) {
      return {};
    }

    return result;
  });
}

export default function ({
  configFile = null,
  policy = [
    {
      allow: "/",
      cleanParam: null,
      crawlDelay: null,
      userAgent: "*",
    },
  ],
  sitemap = null,
  host = null,
} = {}) {
  let options = {
    host,
    policy,
    sitemap,
  };

  return Promise.resolve()
    .then(() =>
      buildConfig(configFile).then((result) => {
        // Need avoid this behaviour in next major release
        // Load config file when it is passed or options were set
        options = Object.assign({}, options, result.config);

        return options;
      })
    )
    .then(
      () =>
        new Promise((resolve) => {
          if (options.policy) {
            if (!Array.isArray(options.policy)) {
              throw new Error("Options `policy` must be array");
            }

            options.policy.forEach((item) => {
              if (!item.userAgent || item.userAgent.length === 0) {
                throw new Error(
                  "Each `policy` should have a single string `userAgent` option"
                );
              }

              if (
                item.crawlDelay &&
                typeof item.crawlDelay !== "number" &&
                !isFinite(item.crawlDelay)
              ) {
                throw new Error(
                  "Option `crawlDelay` must be an integer or a float"
                );
              }

              if (item.cleanParam) {
                if (
                  typeof item.cleanParam === "string" &&
                  item.cleanParam.length > 500
                ) {
                  throw new Error(
                    "Option `cleanParam` should have no more than 500 characters"
                  );
                } else if (Array.isArray(item.cleanParam)) {
                  item.cleanParam.forEach((subItem) => {
                    if (typeof subItem === "string" && subItem.length > 500) {
                      throw new Error(
                        "String in `cleanParam` option should have no more than 500 characters"
                      );
                    } else if (typeof subItem !== "string") {
                      throw new Error(
                        "String in `cleanParam` option should be a string"
                      );
                    }
                  });
                } else if (
                  typeof item.cleanParam !== "string" &&
                  !Array.isArray(item.cleanParam)
                ) {
                  throw new Error(
                    "Option `cleanParam` should be a string or an array"
                  );
                }
              }
            });
          } else {
            throw new Error("Options `policy` should be define");
          }

          if (options.sitemap) {
            if (
              typeof options.sitemap === "string" &&
              !isAbsoluteUrl(options.sitemap)
            ) {
              throw new Error("Option `sitemap` should be an absolute URL");
            } else if (Array.isArray(options.sitemap)) {
              options.sitemap.forEach((item) => {
                if (typeof item === "string" && !isAbsoluteUrl(item)) {
                  throw new Error(
                    "Item in `sitemap` option should be an absolute URL"
                  );
                } else if (typeof item !== "string") {
                  throw new Error(
                    "Item in `sitemap` option should be a string"
                  );
                }
              });
            } else if (
              typeof options.sitemap !== "string" &&
              !Array.isArray(options.sitemap)
            ) {
              throw new Error(
                "Option `sitemap` should be a string or an array"
              );
            }
          }

          if (options.host) {
            if (typeof options.host !== "string") {
              throw new Error("Options `host` must be only one string");
            }

            if (ipRegex({ exact: true }).test(options.host)) {
              throw new Error("Options `host` should be not an IP address");
            }
          }

          let contents = "";

          options.policy.forEach((item, index) => {
            contents += generatePoliceItem(item, index);
          });

          if (options.sitemap) {
            contents += addLine("Sitemap", options.sitemap);
          }

          if (options.host) {
            let normalizeHost = options.host;

            if (normalizeHost.search(/^https?:\/\//) === -1) {
              normalizeHost = `http://${host}`;
            }

            // eslint-disable-next-line node/no-deprecated-api
            const parsedURL = url.parse(normalizeHost, false, true);

            if (!parsedURL.host) {
              throw new Error("Option `host` does not contain correct host");
            }

            let formattedHost = url.format({
              host:
                parsedURL.port && parsedURL.port === "80"
                  ? parsedURL.hostname
                  : parsedURL.host,
              port:
                parsedURL.port && parsedURL.port === "80" ? "" : parsedURL.port,
              protocol: parsedURL.protocol,
            });

            formattedHost = formattedHost.replace(/^http:\/\//, "");

            contents += addLine("Host", formattedHost);
          }

          return resolve(contents);
        })
    );
}
